Adding formatting capabilities to react-codemirror

simonthuresson opened the pull request:
Temporary pull request for Tomas to review

tomasnyberg committed:
Adds standardization check to formatter tests (#334)

tomasnyberg commented on packages/language-support/src/formatting/formatting.ts:
```diff
@@ -45,11 +46,21 @@ export class TreePrintVisitor extends CypherCmdParserVisitor<void> {
   indentation = 0;
   indentationSpaces = 2;
+  targetToken = -1;
+  cursorPos = -1;
```
-1 seems like an invalid value. Is it possible for the editor / formatter to try and find it and throw an error? Can they be initialized to something like 0, 0 instead?

tomasnyberg commented on packages/language-support/src/formatting/formatting.ts:
```diff
@@ -45,11 +46,21 @@ export class TreePrintVisitor extends CypherCmdParserVisitor<void> {
+  setTargetToken = (targetToken: number): void => {
+    this.targetToken = targetToken;
+  };
+
```
are these methods needed? Why not access the field directly?

tomasnyberg commented on packages/language-support/src/formatting/formatting.ts:
```diff
@@ -227,7 +243,11 @@ export class TreePrintVisitor extends CypherCmdParserVisitor<void> {
+      this.cursorPos = this.buffer.join('').length;
+    }
     this.buffer.push(result);
+
```
unnecessary newline

tomasnyberg commented on packages/language-support/src/formatting/formatting.ts:
```diff
@@ -368,13 +380,48 @@ export class TreePrintVisitor extends CypherCmdParserVisitor<void> {
+
+  if (cursorPosition === undefined) {
+    const result = visitor.format(tree);
+    return result;
```
nit: return without saving as a variable

tomasnyberg commented on packages/react-codemirror/src/CypherEditor.tsx:
```diff
@@ -276,6 +276,28 @@ export class CypherEditor extends Component<
   private schemaRef: React.MutableRefObject<CypherConfig> = createRef();
 
+  /**
+   * Format query code
```
dubbelkaka?

tomasnyberg commented on packages/react-codemirror/src/CypherEditor.tsx:
```diff
@@ -276,6 +276,28 @@ export class CypherEditor extends Component<
+  /**
+   * Format query code
+   */
+  format() {
```
if possible: hide completely?

tomasnyberg commented on packages/react-codemirror/src/CypherEditor.tsx:
```diff
@@ -352,6 +374,7 @@ export class CypherEditor extends Component<
       showSignatureTooltipBelow,
       featureFlags: {
         consoleCommands: true,
+        format: true,
```
shuold not always be on

Commit:
Removed initate variable with arbitrary value, undefined instead

tomasnyberg committed:
Adds subquery and parameter handling to formatter (#336)

Commit:
Removed using break, improved robustness

simonthuresson committed:
Adding formatting capabilities to react-codemirror

(cherry picked from commit ec8d6e52f099ec1677a9e764ab1326e618f055e7)

First iteration of cursor position with lexer

(cherry picked from commit 35a70ac83711482aeb43adffaf82e171d03720de)

Handles space at end of line

(cherry picked from commit 88e38eba54308682f5336e3514961499212954df)

Pretty much working cursor position

(cherry picked from commit 19fa9c255cc67f42b56f2044333b9b732db8b6c7)

Improved correctness

(cherry picked from commit 1b312d4d9c34c75ae5f64a8270d4e99c9558a15c)

Fixed function overload for also finding cursos pos when formatting

(cherry picked from commit ced9abd61b59941ce0d060e4b3aa271459273c09)

Removed console log

(cherry picked from commit 3952198467a8900cac8dd94282d2a4f24ee75a87)

Fixed bug with cursor at start

(cherry picked from commit ec09fdb4d28fdd627a91d3ea9f739531d08554f8)

Simplified expression

(cherry picked from commit 374bf84f22488cefec3d82534138588be9c3e1a5)

Improved cursor finder

(cherry picked from commit 88e19bdf5679a1790b86747131d656b84becaf9b)

Extracted function to helper

(cherry picked from commit b7487c2dfc7561bcd0a6195ef7f47b9dd2db8f0c)

Extracted formatter to helper

(cherry picked from commit 4411187a0a4d76d28bd94b81d5f6417b2d7d708f)

Removed unused code

(cherry picked from commit 296ad45d13d0755ccefd6a344f4b7f318a551ea6)

Removed function

(cherry picked from commit e6951b52381019fc27d9a2e185fef0c1dbe5c774)

Uses feature flag to integrate formatter

(cherry picked from commit 9b952e891fda1435c26078790d9044ba3770368d)

Removed always having formatter

(cherry picked from commit 3359d2ffebd60fa28f776d8e3e6ba4de03b43890)

Removed unused import

(cherry picked from commit a56930061c2e9f943295c50d7bfd1edd8413ac3f)

Added format button to codemirror playground

(cherry picked from commit bd1bf958f0a9bc1a62655830a8aed4f9bba4b93e)

Fixed bug when featureFlag was not supplied

(cherry picked from commit a852806b6fefcb0cd0a9b82d95649f833884833c)

simonthuresson committed:
Removed unused functions

simonthuresson committed:
Small fixes

simonthuresson committed:
Removed feature flag always being true

simonthuresson committed:
Fixed issue from merging

simonthuresson committed:
Fixed Typo

simonthuresson committed:
Improved comment on exposed function

Co-authored-by: Oskar Damkjær <OskarDamkjaer@users.noreply.github.com>

simonthuresson committed:
Removed having formatting under a feature flag in the demo

simonthuresson committed:
Added tests for checking new cursos position

simonthuresson committed:
Added tooltip on format button

simonthuresson committed:
Fixed test after merging main

simonthuresson committed:
Merge branch 'main' into add_formatting_keybind_button

simonthuresson committed:
Removed feature flags completely

tomasnyberg commented on packages/language-support/src/tests/formatting/formatting.test.ts:
```diff
@@ -277,3 +277,50 @@ describe('various edgecases', () => {
+  });
+
+  test('cursor start of line without spaces', () => {
+    const query = `MATCH (variable :Label)-[:REL_TYPE]->() 
```
nit: missing space between variable and label, should be `(variable:Label)`
